File: tests/Dali.Tests/TestHarness.cs
using SurrealDb.Embedded.InMemory;

namespace Dali.Tests;

public static class TestHarness
{
    public static IDocumentStore CreateStore()
    {
        var store = Documents.For(o =>
        {
            o.ClientFactory = () => new SurrealDbMemoryClient();
        });

        store.InitializeAsync().GetAwaiter().GetResult();
        return store;
    }

    public static async Task<IDocumentStore> CreateStoreAsync(Action<StoreOptions>? configure = null)
    {
        var store = Documents.For(o =>
        {
            o.ClientFactory = () => new SurrealDbMemoryClient();
            configure?.Invoke(o);
        });

        await store.InitializeAsync();
        return store;
    }
}

File: tests/AeroDB.Sable.FSharp.Tests/TestHarness.fs
namespace AeroDB.Sable.FSharp.Tests

open System
open System.Threading
open System.Threading.Tasks
open AeroDB.Sable
open SurrealDb.Embedded.InMemory
open SurrealDb.Net

[<RequireQualifiedAccess>]
module TestHarness =
    let mutable private counter = 0

    let private uniqueNamespace () =
        let sequence = Interlocked.Increment(&counter)
        let suffix = Guid.NewGuid().ToString("N").Substring(0, 8)
        $"fsharp_test_{sequence}_{suffix}"

    let createStoreAsync () : Task<IDocumentStore> =
        task {
            let namespaceName = uniqueNamespace ()

            let store =
                Documents.For(
                    Action<StoreOptions>(fun options ->
                        options.ClientFactory <-
                            Func<ISurrealDbClient>(fun () -> new SurrealDbMemoryClient())

                        options.Namespace <- namespaceName
                        options.Database <- namespaceName
                        options.Schema.For<TestUser>().SetSchemaMode(SchemaMode.Flexible)
                        |> ignore)
                )

            do! store.InitializeAsync()
            return store
        }
